### ISSUE TITLE
Link role in accessibility tree lacks value attribute

### ISSUE THREAD

**Issue by jamalmazrui** [puppeteer/puppeteer#6311]
I am a screen reader user and developer who values the accessibility tree JSON that can be obtained through the Puppeteer API.  It seems like the "link" role has a name but not value populated.  I think the value should contain the href of the link, but it seems to be empty.

**Comment by christian-bromann** [comment]
@jamalmazrui given this script example:

```js
const puppeteer = require('puppeteer');

(async () => {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();
  await page.goto('https://example.com');
  const snapshot = await page.accessibility.snapshot();
  console.log(snapshot);

  await browser.close();
})();
```

We have 1 link that as you mentioned correctly only has a `role` and `name` property:

```json
{
    "role": "link",
    "name": "More information..."
}
```

The accessibility information from `Accessibility.getFullAXTree` are as follows:

```json
    {
        "nodeId": "18",
        "ignored": false,
        "role": {
            "type": "role",
            "value": "link"
        },
        "name": {
            "type": "computedString",
            "value": "More information...",
            "sources": [
                {
                    "type": "relatedElement",
                    "attribute": "aria-labelledby"
                },
                {
                    "type": "attribute",
                    "attribute": "aria-label"
                },
                {
                    "type": "contents",
                    "value": {
                        "type": "computedString",
                        "value": "More information..."
                    }
                },
                {
                    "type": "attribute",
                    "attribute": "title",
                    "superseded": true
                }
            ]
        },
        "properties": [
            {
                "name": "focusable",
                "value": {
                    "type": "booleanOrUndefined",
                    "value": true
                }
            }
        ],
        "childIds": [
            "19"
        ],
        "backendDOMNodeId": 12
    }
```

What would you suggest the `page.accessibility.snapshot()` should return for such a link?

**Comment by jamalmazrui** [comment]
With multiple browsers, I get a security warning when attempting to open 
the URL https://example.com, so I'm reluctant to examine it directly.  I 
should also mention that I have been using the "interestingly-only" 
option rather than getting the full accessibility tree, in case that 
makes a difference.


In general, I would expect the role of "link" in an accessibility tree 
to have the the clickable text as the "name" attribute and the 
underlying URL as the "value" attribute.  For example, this is how the 
NVDA screen reader (free, open source on Windows) represents a link in 
its accessibility tree.


n 8/6/2020 7:49 AM, Christian Bromann wrote:

**Comment by christian-bromann** [comment]
> so I'm reluctant to examine it directly

Interesting, I don't get this warning. This is how the DOM looks like:

![Screenshot 2020-08-06 at 18 52 54](https://user-images.githubusercontent.com/731337/89559581-25022880-d816-11ea-9289-e3c232deb170.png)

So if I understand you correctly you suggest that the accessibility tree item would look as follows:

```json
{
    "role": "link",
    "name": "More information...",
    "value": "https://www.iana.org/domains/example"
}
```

If this is ok for the Puppeteer team, I would be happy to raise a PR.

**Comment by jamalmazrui** [comment]
Yes, I support that proposal, thanks.

**Comment by nicolashenry** [comment]
Not sure if this is related but I had to activate accessibility first to have the full accessibility resolved by accessibility.snapshot():
```js
const page = await browser.newPage();
await page._client().send('Accessibility.enable');
```

**Comment by VoidMonk** [comment]
Just to chime in since this issue is old but important.

The accessibility tree item for a link still doesn't contain its URL value:

```js
await page.accessibility.snapshot({ interestingOnly: false });
```

..it might return a link as..

```json
{
    "role": "link",
    "name": "More information...",
}
```

..when it should also include the link URL as value for completeness and usability..

```json
{
    "role": "link",
    "name": "More information...",
    "value": "https://www.iana.org/domains/example"
}
```

**Comment by jeff-an** [comment]
Also encountering this problem w/ checkbox and input names not populating, e.g.
```
  {
    nodeId: '1245',
    ignored: false,
    role: { type: 'role', value: 'checkbox' },
    chromeRole: { type: 'internalRole', value: 14 },
    name: { type: 'computedString', value: '', sources: [Array] },
    properties: [ [Object], [Object], [Object] ],
    parentId: '1242',
    childIds: [],
    backendDOMNodeId: 1424
  }
``` 

could the core of the problem be that `Accessibility.enable` returns too quickly w/out fully completing?

an easy repro is looking at the checkboxes on [booking.com](https://booking.com)

**Comment by OrKoN** [comment]
So the link value are not being sent by Chrome. I see that there are [only two situations](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/accessibility/inspector_accessibility_agent.cc;l=971;drc=aef1f146e39ffed5aa699d60e6d0f6e9c6515c24) One covers range values and the other gets value for controls. And it looks like the link is not considered a control.

**Comment by aleventhal** [comment]
Hi Jamal! Having the value attribute filled with the URL is a legacy windows MSAA thing, we do that in the downstream windows layer, similar to how the MSAA value of a document is the url. The value in our cross-platform Blink layer only represents user input in a form control. However, there is a more complete answer to this question, since it seems reasonable to want to get the url field for a link, which Blink puts in a property called url.

I don't know much about Puppeteer, but it sounds like it uses the CDP API here: https://chromedevtools.github.io/devtools-protocol/tot/Accessibility/#method-getFullAXTree
Unfortunately, CDP does not seem to map all of the useful properties provided by Blink in an AXNode (property fields are listed at ax_enums.mojom in the Chromium source tree), and does not map the url property afaict.

**Comment by OrKoN** [comment]
I have added the URL property in https://chromium-review.googlesource.com/c/chromium/src/+/5631015 Once it is in stable, we can update Puppeteer and add a test.
